torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -947,13 +969,13 @@ Tensor native_dropout_double_backward(const Tensor& ggI, const Tensor& grad, con
 }
 
 Tensor evenly_distribute_backward(Tensor grad, const Tensor & input, const Tensor & value) {
-  if (input.is_cuda()) {
-    auto mask = (input == value).logical_or_(input.isnan().logical_and_(value.isnan()));
-    return mask * (grad / mask.sum());
-  } else {
+  if (!areAnyTensorSubclassLike({grad, input, value}) && input.is_cpu()) {
+    // fast-path on CPU
     auto mask = value.isnan().item<bool>() ? input.isnan() : input == value;
     return grad.new_zeros(input.sizes(), input.options()).masked_fill_(mask, grad / mask.sum());
   }
+  auto mask = (input == value).logical_or_(input.isnan().logical_and_(value.isnan()));
```

Comment:
this is also doing inplace. That's fine for you?

Reply:
Yes. Let's assume that input is a subclass A. Then `value` is also a subclass A (because value is the result of the operation in the forward pass). We are able to do (A == A).inplace_(A.inplace_(A)).

Reply:
On second thought, this works for vmap, but I don't know if `A.inplace_(A)` works for all Tensor subclasses

Reply:
Proposal: a subclass A that does not support (a: A).inplace_(b: A) in cases where regular Tensors do should rely on always-on functionalization to solve their problem

Reply:
Note that the `input.isnan().logical_and_(...` does still sounds problematic if value is batched but not input?

Reply:
Updated